### ISSUE TITLE
Type instability in @ein

### ISSUE THREAD

**Issue by jaemolihm** [under-Peter/OMEinsum.jl#151]
I tried the same example from #92 and found type instability.

```julia
julia> using OMEinsum

julia> f1(x, y) = @ein z[i,k,l] := x[i,j] * y[j,k,l]

f1 (generic function with 1 method)

julia> x = rand(3, 5);

julia> y = rand(5, 11, 7);

julia> @code_warntype f1(x, y)
MethodInstance for f1(::Matrix{Float64}, ::Array{Float64, 3})
  from f1(x, y) in Main at REPL[2]:1
Arguments
  #self#::Core.Const(f1)
  x::Matrix{Float64}
  y::Array{Float64, 3}
Locals
  z::Any
Body::Any
1 ─ %1 = Core.tuple((1, 2), (2, 3, 4))::Core.Const(((1, 2), (2, 3, 4)))
│   %2 = OMEinsum.EinCode(%1, (1, 3, 4))::DynamicEinCode{Int64}
│   %3 = Core.tuple(x, y)::Tuple{Matrix{Float64}, Array{Float64, 3}}
│   %4 = OMEinsum.einsum(%2, %3)::Any
│        (z = %4)
└──      return %4


julia> versioninfo()
Julia Version 1.8.2
Commit 36034abf260 (2022-09-29 15:21 UTC)
Platform Info:
  OS: Linux (x86_64-linux-gnu)
  CPU: 32 × Intel(R) Xeon(R) CPU E5-2630 v3 @ 2.40GHz
  WORD_SIZE: 64
  LIBM: libopenlibm
  LLVM: libLLVM-13.0.1 (ORCJIT, haswell)
  Threads: 1 on 32 virtual cores
Environment:

(vertex) pkg> status OMEinsum
Status `~/vertex/Project.toml`
  [ebe7aa44] OMEinsum v0.7.2
Warning The project dependencies or compat requirements have changed since the manifest was last resolved. It is recommended to `Pkg.resolve()` or consider `Pkg.update()` if necessary.
```

**Comment by GiggleLiu** [comment]
Thanks for the report of type instability, it will be fixed in PR #152 .

**Comment by GiggleLiu** [comment]
Since making einsum type stable requires using generated functions, which can significantly slow down the compiling process. OMEinsum is not design to be type stable.
Wondering what is the problem that you want to solve by making the code type stable? If you want to prevent the code becoming unstable, you can add a type annotation to your function.
